rLapz/server.c

```diff
@@ -1,7 +1,198 @@
+	int socket_option			= 1;
+	int server_fd				= 0;
+	int client_fd				= 0;
+	struct sockaddr_in server	= {0};
+	struct sockaddr_in client	= {0};
+	socklen_t client_len		= 0;
+
+	/* Create socket */
+	server_fd = socket(AF_INET, SOCK_STREAM, 0);
+	if (server_fd < 0) {
+		perror("Create socket");
+		is_error = TRUE;
+		goto cleanup;
+	}
+
+	/* Socket option
+	 * Reuse IP address */
+	if (setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR,
+					&socket_option, sizeof(socket_option)) < 0) {
+		perror("Reuse address");
+		is_error = TRUE;
+		goto cleanup;
+	}
+
+	/* TCP configuration */
+	server.sin_family = AF_INET;
+	server.sin_addr.s_addr = inet_addr(addr);
+	server.sin_port = htons(port);
+
+	/* binding server address */
+	if (bind(server_fd, (struct sockaddr *)&server, sizeof(server)) < 0) {
+		perror("Bind socket");
+		is_error = TRUE;
+		goto cleanup;
+	}
+
+	/* listening... */
+	if (listen(server_fd, 3) < 0) {
+		perror("Listening");
+		is_error = TRUE;
+		goto cleanup;
+	}
+	printf("Listening: [%s:%d]...\n", addr, port);
+
+	/* get file properties from client */
+	while (1) {
+		client_len = sizeof(struct sockaddr_in);
+		puts("Waiting for incoming connection...");
+		client_fd = accept(server_fd, (struct sockaddr *)&client, &client_len);
+		if (client_fd < 0) {
+			perror("Accepting connection");
+			continue;
+		} else {
+			puts("Connection accepted");
+			file_handler(&client_fd, &server_fd);
+		}
+	}
+
+cleanup:
+	close(server_fd);
+	close(client_fd);
+	if (is_error == TRUE)
+		return EXIT_FAILURE;
+
+	return EXIT_SUCCESS;
+}
+
+static void
+file_handler(int *client_fd, int *server_fd)
+{
+	ssize_t recv_bytes		= 0;
+	ssize_t recv_ok_bytes	= 0;
+	ssize_t write_bytes		= 0;
+	size_t total_bytes		= 0;
+
+	short int is_error		= 0;
+	int file_desc			= 0;
+	char target_file[255]	= {0};
+	/* why heap? because I need flexible memory size, though. */
+	char *data_arena		= calloc(sizeof(packet), sizeof(packet));
+	packet *packet_data		= (packet*)data_arena;
+	
+	/* get file properties */
+	while (recv_ok_bytes < sizeof(packet)) {
+		recv_bytes = recv(*client_fd, (char*)packet_data+recv_ok_bytes, BUFFER_SIZE, 0);
+		if (recv_bytes < 0) {
+			perror("Recv");
+			goto cleanup;
+		}
+		recv_ok_bytes += recv_bytes;
+	}
+
+	/* fill packet_data->filename with file properies that received from client */
+	snprintf(target_file, sizeof(packet_data->filename),
+			"uploaded_files/%s", packet_data->filename);
+
+	/* you know what is these :3 */
+	puts("\nFile info: ");
+	printf(" -> File name\t\t: %s\n", packet_data->filename);
+	printf(" -> File name length\t: %d\n", packet_data->filename_len);
+	printf(" -> File size\t\t: %zu byte\n", packet_data->file_size);
+	printf(" -> Stored at\t\t: %s\n", target_file);
+
+	/* reallocation memory size */
+	/* I know this kinda not safe */
+	data_arena = realloc(data_arena, sizeof(packet) + packet_data->file_size);
+	memset(data_arena, 0, sizeof(data_arena));
+	packet_data = (packet*)data_arena;
+
```

Comment:
```c
memset(data_arena, 0, sizeof(data_arena)); // Why sizeof(data_arena)?
```

You probably want to write
```c
memset(data_arena, 0, sizeof(packet) + packet_data->file_size);
```

Reply:
Oh no!
You probably want to get rid of `memset`.

Why did you `memset` it anyway?
Doesn't it clobber the `filename_len` and `filename`?

Reply:
I want to "clear" whole data inside `packet_data` including `filename_len` and `filename`, but I seems wrong.

Reply:
wait, I forget "backup" `filename_len` and `filename` data to new variable,